docs/MATCHMAKING_PROPOSAL.md

```diff
@@ -0,0 +1,171 @@
+This proposal details the specifications for the matchmaking in our app.
+## Overview
+### 1. Queue
+
+The user presses the PLAY button, making him join a queue for the chosen gamemode.
+
+To indicate the server we are queueing for a game, we call the HTTPS endpoint `/queue` with the following parameters:
+
+* `game_mode`: the game mode we are queueing for
+
+*Note: We do not need to provide the user id, as HTTPS cloud functions authenticate the user with their already authenticated firebase identity, which allows to obtain the user uid without explicitly providing it. It also allows to have some security as only authenticated users can call this function. [Source](https://firebase.google.com/docs/functions/callable)*
+### 2. The game - server-side
+
+Once the server receives a request to queue, it acts as follows:
+
+1. Looks for games that are not full and have the same game mode.
+2. If there are no such games, creates a new game with the same game mode.
+3. Adds the user to the game.
+4. If the game is full, starts the game.
+5. If the game is not full, waits for the other players to join.
+6. Either way, returns the game id to the user.
+
+The game document `/game/{gameId}` looks like this:
+```jsonc
+// game.json
+
+{
+  "mode": "5P,PC,3R,0T",
+  "timestamp": "2020-01-01T00:00:00.000Z",
+  "rounds": [],
```

Comment:
rounds_ids?

Reply:
We could, but on firebase you almost always use IDs, almost never actually nested maps, so this feels redundant.